### PR TITLE
Run rails before_eager_load hooks in resque setup

### PR DESCRIPTION
**Pull request by abhi-patel**
Rails has added support for running hooks before eager load. Resque should also be executing the same hooks before calling `Rails.application.eager_load!`

### CODE REVIEW

**Comment by abhi-patel** [comment]
Checkout https://github.com/rails/rails/blob/master/railties/lib/rails/application/finisher.rb#L58

Rails explicitly runs before_eager_load hooks while eager loading. Rescue is missing the same and hence we ran into some issues where we tried running some code when eager load is enabled. 

Everything worked fine for normal code while the same code was throwing exceptions for Resque workers.

aws/aws-sdk-ruby#1229

**Comment by fw42** [comment]
@rafaelfranca, does this make sense to you?

**Comment by rafaelfranca** [comment]
Yes it does. But this code is not eager loading engines for example so this kind of issues can still happen when using an engine. I'll comment inline what can be done but it will not work for Rails 3.

File: lib/resque/tasks.rb [hunk 1]
```diff
@@ -41,6 +41,7 @@
   task :preload => :setup do
     if defined?(Rails) && Rails.respond_to?(:application)
       # Rails 3
+      ActiveSupport.run_load_hooks(:before_eager_load, Rails.application)
       Rails.application.eager_load!
```
Comment (rafaelfranca) [inline lib/resque/tasks.rb]: For Rails 4 it should be `Rails.application.config.eager_load_namespaces.each(&:eager_load!)`

**Comment by abhi-patel** [comment]
@fw42 - Would It be fine to add condition here based on Rails version? Or check if it responds to 'eager_load_namespaces' method?

File: lib/resque/tasks.rb (hunk 1 above)
Comment (abhi-patel) [inline lib/resque/tasks.rb]: `Rails.application.config.eager_load_namespaces.each(&:eager_load!)` will run eager_load for specified classes/engines.

I suppose will still have to call `ActiveSupport.run_load_hooks(:before_eager_load, Rails.application)` to run the hooks, Or will they execute automatically(internally) by above line(eager_load!)?

Comment (rafaelfranca) [inline lib/resque/tasks.rb]: You are still supposed to call that line to run the hooks.

**Comment by abhi-patel** [comment]
Look like builds failed because of gemfile isssue. Should be fine now.

**Comment by abhi-patel** [comment]
@fw42 - Any update on this?

**Comment by fw42** [comment]
Can you rebase your branch? I think those CI failures have been fixed on master.

As for the actual changes: I don't feel qualified to review this, but if @rafaelfranca thinks that the changes here make sense and if one other person from resque can +1 this, then I'm happy to merge.

@steveklabnik?

**Comment by steveklabnik** [comment]
So, this looks fine to me, but I will admit that I'm a bit out of the loop on exactly what's required to run when here. If @rafaelfranca says :+1: I'm :+1:

**Comment by abhi-patel** [comment]
Just Rebased. Its still broken because of some other gem issue.

**Comment by fw42** [comment]
@rafaelfranca does this look good to merge to you?

**Comment by rafaelfranca** [comment]
:shipit:

**Pull request merged**